app/models/minion.rb

```diff
@@ -4,18 +4,43 @@
 
 # Minion represents the minions that have been registered in this application.
 class Minion < ApplicationRecord
+  class NotEnoughMinions < StandardError; end
+  class CouldNotAssignRole < StandardError; end
+
   enum role: [:master, :minion]
 
   validates :hostname, uniqueness: true
 
+  # This method is used to assign the specified roles to Minions which do not
+  # have a roles already assigned (available minions).
+  # roles param are the roles we absolutely have to assign. If we can't assign
+  # one of those, the method will raise.
+  # default_role param can be set if we want all the rest of the available
+  # minions to get a default role.
+  def self.assign_roles(roles: [], default_role: nil)
+    minions = Minion.where(role: nil)
+
+    # only the needed number of minions or all if we have a default role
+    minions = minions.limit(roles.size) unless default_role
```

Comment:
Dumb question: won't this produce a second SQL query ? If so, could you simply chain this method call with the previous one for this case ?